build/utils/yarn.go

```diff
@@ -91,26 +93,105 @@ func GetYarnExecutable() (string, error) {
+	if err != nil {
+		return
+	}
+
+	isV2AndAbove := version.NewVersion(executableVersionStr).Compare(yarnV2Version) <= 0
+
+	// Run 'yarn info or list'
+	responseStr, errStr, err := runYarnInfoOrList(executablePath, srcPath, isV2AndAbove)
 	// Some warnings and messages of Yarn are printed to stderr. They don't necessarily cause the command to fail, but we'd want to show them to the user.
 	if len(errStr) > 0 {
 		log.Warn("An error occurred while collecting dependencies info:\n" + errStr)
 	}
 	if err != nil {
 		log.Warn("An error was thrown while collecting dependencies info: " + err.Error() + "\nCommand output:\n" + responseStr)
 
+		// Spacial case: when 'yarn install' wasn't executed on the project we will get an error with non-empty responseStr. for yarn v2 and v3 ONLY.
+		if strings.Contains(responseStr, "present in your lockfile") {
+			err = errors.New("fetching dependencies failed since '" + packageInfo.Name + "' doesn't present in your lockfile\nPlease run 'yarn install' to update lockfile\n" + err.Error())
+			return
+		}
+
 		// A returned error doesn't necessarily mean that the operation totally failed. If, in addition, the response is empty, then it probably failed.
 		if responseStr == "" {
 			return
 		}
-		err = nil
 	}
 
+	if isV2AndAbove {
+		dependenciesMap, root, err = buildYarnV2DependencyMap(packageInfo, responseStr)
+	} else {
+		dependenciesMap, root, err = buildYarnV1DependencyMap(packageInfo, responseStr)
+	}
+	return
+}
+
+// getVersion gets the current project's yarn version
+func GetVersion(executablePath, srcPath string) (string, error) {
+	command := exec.Command(executablePath, "--version")
+	command.Dir = srcPath
+	outBuffer := bytes.NewBuffer([]byte{})
+	command.Stdout = outBuffer
+	errBuffer := bytes.NewBuffer([]byte{})
+	command.Stderr = errBuffer
+	err := command.Run()
+	if _, ok := err.(*exec.ExitError); ok {
+		err = errors.New("An error occurred while attempting to run the 'yarn --version' command. The executed yarn version could not be retrieved:\n" + err.Error())
+	}
+	return strings.TrimSpace(outBuffer.String()), err
+}
+
+// buildYarnV1DependencyMap builds a map of dependencies for Yarn versions < 2.0.0
+// Pay attention that in Yarn < 2.0.0 the project itself with its direct dependencies is not presented when running the
+// command 'yarn list' therefore the root is built manually.
+func buildYarnV1DependencyMap(packageInfo *PackageInfo, responseStr string) (dependenciesMap map[string]*YarnDependency, root *YarnDependency, err error) {
+	dependenciesMap = make(map[string]*YarnDependency)
+	var depTree Yarn1Data
+	err = json.Unmarshal([]byte(responseStr), &depTree)
+	if err != nil {
+		err = errors.New("Couldn't parse 'yarn list' results in order to create the dependencyMap:\n" + err.Error())
+		return
+	}
+
+	if depTree.Data.DepTree == nil {
+		err = errors.New("Error: (buildYarnV1DependencyMap) depTree struct received a nil value in a required field: depTree.Data.DepTree")
```

Comment:
should start with lowercase
no need to write `error`
the error is not user friednly